File: opcua-webservice-spring/tutorials/lesson2/src/main/java/org/opcfoundation/uawebservicedemo/models/company/department/DepartmentType.java
package org.opcfoundation.uawebservicedemo.models.company.department;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.opcfoundation.uawebservicedemo.database.EmployeeDatabase;
import org.opcfoundation.uawebservicedemo.database.entity.Department;
import org.opcfoundation.uawebservicedemo.database.mapper.DepartmentMapper;
import org.opcfoundation.uawebservicedemo.models.EmployeeTwinSpace;
import org.opcfoundation.webserver.digitaltwin.element.ElementCollectionType;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDescriptorRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDescriptorResponse;

import java.util.concurrent.CompletableFuture;

// Defining department including child department and employee
public class DepartmentType extends ElementCollectionType {
    public DepartmentType(
            DepartmentEmployeeReferenceType departmentEmployeeReferenceType,
            EmployeeTwinSpace twinSpace) {
        super("DepartmentType", new LocalizedText("Department"), twinSpace);

        addReferenceElement(
                departmentEmployeeReferenceType,
                "Employees",
                new LocalizedText("Employees"),
                new LocalizedText("Employees Of department"),
                true);
    }

    public CompletableFuture<GetDescriptorResponse> onGetDescriptor(GetDescriptorRequest request)
    {
        try
        {
            // Get department id from request
            Integer departmentId = Integer.parseInt(request.getId());

            // Database query
            DepartmentMapper departmentMapper = EmployeeDatabase.getDepartmentMapper();
            Department company = departmentMapper.getDepartment(departmentId);

            GetDescriptorResponse response = new GetDescriptorResponse(
                    new LocalizedText(company.getName()),
                    new LocalizedText(company.getComment()));

            return CompletableFuture.completedFuture(response);
        } catch (Exception e) {
            throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);
        }
    }
}

File: opcua-webservice-node/examples/demoserver/src/models/company/department/DepartmentType.ts
import { StatusCodes } from "opcua-webapi";
import {
    UaError,
    UaLocalizedText,
} from "opcua-webapi-ts";
import {
    ElementCollectionType,
    GetDescriptorRequest,
    GetDescriptorResponse,
} from "opcua-webservice-node";

import { prisma } from "../../../connectors/prismaClient";
import { EmployeeTwinSpace } from "../../EmployeeTwinSpace";
import type { DepartmentEmployeeReferenceType } from "./DepartmentEmployeeReferenceType";

export class DepartmentType extends ElementCollectionType {
    constructor(
            departmentEmployeeReferenceType: DepartmentEmployeeReferenceType,
            twinSpace: EmployeeTwinSpace) {
        super("DepartmentType", new UaLocalizedText("Department"), twinSpace);

        this.addReferenceElement(
                departmentEmployeeReferenceType,
                "Employees",
                new UaLocalizedText("Employees"),
                new UaLocalizedText("Employees Of department"),
                true);
    }

    async onGetDescriptor(request : GetDescriptorRequest) : Promise<GetDescriptorResponse>
    {
        // Get department id from request
        const departmentId = Number.parseInt(request.id, 10);
        if (Number.isNaN(departmentId)) {
            throw UaError.from(StatusCodes.BadNodeIdUnknown);
        }

        // Database query
        const department = await prisma.department.findUnique({
            where: { ID: departmentId },
            select: { Name: true, Comment: true },
        });

        if (department === null) {
            throw UaError.from(StatusCodes.BadNodeIdUnknown);
        }

        return new GetDescriptorResponse(
            new UaLocalizedText(department.Name),
            new UaLocalizedText(department.Comment ?? ""));
    }
}